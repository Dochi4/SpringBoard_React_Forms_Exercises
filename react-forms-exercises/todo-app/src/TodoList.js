import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import NewTodo from "./NewTodo";
import EditTodo from "./EditTodo";

const TodoList = () => {
  const initialState = [
    { id: uuid(), item: "todo item 1" },
    { id: uuid(), item: "todo item 2" },
  ];

  const [todos, setTodos] = useState(initialState);

  const addTodo = (newTodoData) => {
    setTodos((todos) => [...todos, { ...newTodoData, id: uuid() }]);
  };
  const handleRemoveItem = (idToRemove) => {
    setTodos((todos) => todos.filter((todo) => todo.id !== idToRemove));
  };

  const toggleTodoCompleted = (idToToggle) => {
    setTodos((todos) =>
      todos.map((todo) =>
        todo.id === idToToggle ? { ...todo, completed: !todo.completed } : todo
      )
    );
  };

  function editTodo(id, newText) {
    const edited = todos.map((todo) =>
      todo.id === id ? { ...todo, item: newText } : todo
    );
    setTodos(edited);
  }

  return (
    <div>
      <div>
        <NewTodo addTodo={addTodo} />
      </div>
      <ul className="container">
        {todos.map((todo) => (
          <EditTodo
            key={todo.id}
            id={todo.id}
            item={todo.item}
            completed={todo.completed}
            toggleTaskCompleted={toggleTodoCompleted}
            editTodo={editTodo}
            removeTodo={handleRemoveItem}
          />
        ))}
      </ul>
    </div>
  );
};

export default TodoList;
