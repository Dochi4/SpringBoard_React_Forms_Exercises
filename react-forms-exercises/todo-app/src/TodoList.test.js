import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import TEST_TODO from "./_testCommon";
import TodoList from "./TodoList";

it("should render without crashing", function () {
  render(<TodoList />);
});

it("matches snapshot", function () {
  const { asFragment, container } = render(<TodoList />);
  container.querySelectorAll("[id]").forEach((el) => el.removeAttribute("id"));
  container
    .querySelectorAll("[for]")
    .forEach((el) => el.removeAttribute("for"));
  expect(asFragment()).toMatchSnapshot();
});
it("should add new Todo", function () {
  const { getByLabelText, getByText, getAllByText } = render(<TodoList />);

  const initialCount = getAllByText("Remove").length;

  fireEvent.change(getByLabelText("Item"), {
    target: { value: TEST_TODO.item },
  });
  fireEvent.click(getByText("Add Todo"));

  expect(getAllByText("Remove").length).toBe(initialCount + 1);
});

it("should not add Todo", function () {
  const { getByLabelText, getByText, getAllByText } = render(<TodoList />);

  fireEvent.change(getByLabelText("Item"), {
    target: { value: "" },
  });

  fireEvent.click(getByText("Add Todo"));

  expect(getAllByText("Remove").length).toBe(2);
});

it("should edit an existing Todo", function () {
  const { getAllByText, getByDisplayValue, getByText } = render(<TodoList />);

  fireEvent.click(getAllByText("Edit")[0]);

  const itemInput = getByDisplayValue("todo item 1");
  fireEvent.change(itemInput, { target: { value: "Newitem" } });

  fireEvent.click(getByText("Save"));
  expect(getByText("Newitem")).toBeInTheDocument();
});

it("should test Remove", function () {
  const { getAllByText } = render(<TodoList />);

  expect(getAllByText("Remove").length).toBe(2);

  fireEvent.click(getAllByText("Remove")[0]);

  expect(getAllByText("Remove").length).toBe(1);
});
