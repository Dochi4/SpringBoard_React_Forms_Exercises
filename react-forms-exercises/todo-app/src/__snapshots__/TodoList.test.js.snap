// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <label>
          Item
        </label>
        <input
          name="item"
          placeholder="Add Todo Item"
          type="text"
          value=""
        />
        <button>
          Add Todo
        </button>
      </form>
    </div>
    <ul
      class="container"
    >
      <li>
        <input
          type="checkbox"
        />
        <label>
          todo item 1
        </label>
        <button
          type="button"
        >
          Edit
        </button>
        <button
          type="button"
        >
          Remove
        </button>
      </li>
      <li>
        <input
          type="checkbox"
        />
        <label>
          todo item 2
        </label>
        <button
          type="button"
        >
          Edit
        </button>
        <button
          type="button"
        >
          Remove
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
